### PR TITLE
feat: add notification provider `SendGrid`

### PR DESCRIPTION
**Pull request by sctnightcore**
⚠️⚠️⚠️ Since we do not accept all types of pull requests and do not want to waste your time. Please be sure that you have read pull request rules:
https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md#can-i-create-a-pull-request-for-uptime-kuma

Tick the checkbox if you understand [x]:
- [x] I have read and understand the pull request rules.

# Description
Screenshot 2567-10-16 at 17 24 53

Fixes #(issue)

## Type of change

Please delete any options that are not relevant.

- Bug fix (non-breaking change which fixes an issue)
- User interface (UI)
- New feature (non-breaking change which adds functionality)
- Breaking change (a fix or feature that would cause existing functionality to not work as expected)
- Other
- This change requires a documentation update

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I ran ESLint and other linters for modified files
- [x] I have performed a self-review of my own code and tested it
- [x] I have commented my code, particularly in hard-to-understand areas (including JSDoc for methods)
- [x] My changes generates no new warnings
- [x] My code needed automated testing. I have added them (this is optional task)

## Screenshots (if any)

Please do not use any external image service. Instead, just paste in or drag and drop the image here, and it will be uploaded automatically.

### CODE REVIEW

File: src/components/notifications/SendGrid.vue [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    </div>
+    <div class="mb-3">
+      <label for="sendgrid-bcc-email" class="form-label">{{ $t("BCC Email") }}</label>
+      <input id="sendgrid-bcc-email" v-model="$parent.notification.sendgridBccEmail" type="email" class="form-control">
+      <small class="form-text text-muted">{{ $t("Separate multiple email addresses with commas") }}</small>
+    </div>
+    <div class="mb-3">
+      <label for="sendgrid-subject" class="form-label">{{ $t("Email Subject") }}</label>
+      <input id="sendgrid-subject" v-model="$parent.notification.sendgridSubject" type="text" class="form-control" required>
+      <small class="form-text text-muted">{{ $t("Default subject will be used if left empty") }}</small>
```
Comment (CommanderStorm) [inline src/components/notifications/SendGrid.vue:28]: The input is marked as `required`, which conflicts with `if left empty`..

File: src/components/notifications/SendGrid.vue [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    </div>
+    <div class="mb-3">
+      <i18n-t tag="p" keypath="More info on:" style="margin-top: 8px;">
+        <a href="https://docs.sendgrid.com/api-reference/mail-send/mail-send" target="_blank">https://docs.sendgrid.com/api-reference/mail-send/mail-send</a>
+      </i18n-t>
+    </div>
```
Comment (CommanderStorm) [inline src/components/notifications/SendGrid.vue:34]: Lets do this do this way:

```suggestion
    <i18n-t tag="p" keypath="More info on:" style="margin-top: 8px;">
      <a href="https://docs.sendgrid.com/api-reference/mail-send/mail-send" target="_blank">https://docs.sendgrid.com/api-reference/mail-send/mail-send</a>
    </i18n-t>
```

File: src/components/notifications/SendGrid.vue [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  </template>
```
Comment (CommanderStorm) [inline src/components/notifications/SendGrid.vue:35]: nit:

```suggestion
  </template>

```

File: src/components/notifications/index.js [hunk 4]
```diff
@@ -139,7 +140,8 @@ const NotificationFormList = {
     "whapi": Whapi,
     "gtxmessaging": GtxMessaging,
     "Cellsynt": Cellsynt,
-    "WPush": WPush
+    "WPush": WPush,
+    "SendGrid": SendGrid
```
Comment (CommanderStorm) [inline src/components/notifications/index.js:144]: nit:

```suggestion
    "SendGrid": SendGrid,
```

File: src/components/notifications/SendGrid.vue [hunk 5]
```diff
@@ -0,0 +1,35 @@
+<template>    
+    <!-- SendGrid fields -->
```
Comment (CommanderStorm) [inline src/components/notifications/SendGrid.vue:2]: nit:

```suggestion
```

File: src/components/NotificationDialog.vue [hunk 6]
```diff
@@ -165,6 +165,7 @@ export default {
                 "whapi": "WhatsApp (Whapi)",
                 "gtxmessaging": "GtxMessaging",
                 "Cellsynt": "Cellsynt",
+                "SendGrid": "SendGrid (Email)"
```
Comment (CommanderStorm) [inline src/components/NotificationDialog.vue:168]: What do you mean by `(Email)` here?
I don't think this is quite nessesary, right?

```suggestion
                "SendGrid": "SendGrid"
```

File: server/notification-providers/send-grid.js [hunk 7]
```diff
@@ -0,0 +1,65 @@
+const NotificationProvider = require("./notification-provider");
+const axios = require("axios");
+
+class SendGrid extends NotificationProvider {
+  name = "SendGrid";
```
Comment (CommanderStorm) [inline server/notification-providers/send-grid.js:5]: please fix the lining issues in this file

File: server/notification-providers/send-grid.js [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+      // Add BCC recipients if provided
+      if (notification.sendgridBccEmail) {
+        personalizations.bcc = notification.sendgridBccEmail
+          .split(",")
+          .map((email) => ({ email: email.trim() }));
+      }
+
+      let data = {
+        personalizations: [personalizations],
+        from: { email: notification.sendgridFromEmail },
```
Comment (CommanderStorm) [inline server/notification-providers/send-grid.js:41]: lets trim this as well

```suggestion
        from: { email: notification.sendgridFromEmail.trim() },
```

File: server/notification-providers/send-grid.js [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        subject:
+          notification.sendgridSubject ||
```
Comment (CommanderStorm) [inline server/notification-providers/send-grid.js:43]: Please make the default more clear to our users by setting it like this instead.

(Read: leave this code, but also add a default in the frontend)

Here is an example how you do this:
https://github.com/louislam/uptime-kuma/blob/85dfe1f5d17dad5e035eb3620b8f516b2bcf79cd/src/components/notifications/TechulusPush.vue#L66-L69

**Review by CommanderStorm — CHANGES_REQUESTED** [review]
Overall, this mostly looks good.
I have left a few requests for changes.

File: server/notification-providers/send-grid.js (hunk 7 above)
Comment (sctnightcore) [reply server/notification-providers/send-grid.js]: done

File: server/notification-providers/send-grid.js (hunk 9 above)
Comment (sctnightcore) [reply server/notification-providers/send-grid.js:43]: done [5f96e67](https://github.com/louislam/uptime-kuma/pull/5205/commits/5f96e6700ed45125cedb3d96051eba773354a741)

File: src/components/notifications/SendGrid.vue (hunk 1 above)
Comment (sctnightcore) [reply src/components/notifications/SendGrid.vue]: done [8b015a6](https://github.com/louislam/uptime-kuma/pull/5205/commits/8b015a6c4a021e8062ccb5cd0ffc379ac9c759d0)

File: src/components/notifications/SendGrid.vue [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+    <div class="mb-3">
+      <label for="sendgrid-api-key" class="form-label">{{ $t("SendGrid API Key") }}</label>
```
[the remark this answers is not in the document]
Comment (sctnightcore) [reply src/components/notifications/SendGrid.vue]: done 
f8101c7

File: src/components/notifications/SendGrid.vue [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+<script>
+import HiddenInput from "../HiddenInput.vue";
+
+export default {
+    components: {
+        HiddenInput,
+    },
+    mounted() {
+        if (typeof this.$parent.notification.sendgridSubject === "undefined") {
+            this.$parent.notification.sendgridSubject = "Uptime-Kuma";
```
Comment (CommanderStorm) [inline src/components/notifications/SendGrid.vue:43]: The backend has the following as a default. We schould only have one default. Which one is better?

```suggestion
            this.$parent.notification.sendgridSubject = "Notification from Your Uptime Kuma";
```

File: src/components/notifications/SendGrid.vue [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    </div>
+    <div class="mb-3">
+        <label for="sendgrid-cc-email" class="form-label">{{ $t("smtpCC") }}</label>
+        <input id="sendgrid-cc-email" v-model="$parent.notification.sendgridCcEmail" type="email" class="form-control">
+        <div class="form-text">{{ $t("Separate multiple email addresses with commas") }}</div>
+    </div>
+    <div class="mb-3">
+        <label for="sendgrid-bcc-email" class="form-label">{{ $t("smtpBCC") }}</label>
+        <input id="sendgrid-bcc-email" v-model="$parent.notification.sendgridBccEmail" type="email" class="form-control">
+        <small class="form-text text-muted">{{ $t("Separate multiple email addresses with commas") }}</small>
```
Comment (CommanderStorm) [inline src/components/notifications/SendGrid.vue:22]: Please add this key as well to `en.json`

image

Comment (sctnightcore) [reply src/components/notifications/SendGrid.vue:22]: done 
0433802

**Comment by sctnightcore** [comment]
> From the code side, this would be a merge. I missed asking you for the screenshots in the description. We require screenshots to ensure that there are no glaring bugs and because testing notification providers do not scale.
> 
> => Could you send a test/up/down or cert-expiry notification via this monitor post a screenshot of it?

I've added a photo to the description.
Thank you for the review.

**Review by CommanderStorm — APPROVED** [review]
Thanks for the new notification provider! 🎉 



> [!NOTE]
> This PR is part of the [`v2.0` merge window](https://github.com/louislam/uptime-kuma/milestone/24) => see #4500 for the bugs that need to be addressed before we can ship this release ^^

<sub>All changes in this PR are small and uncontroversial ⇒ merging with junior maintainer approval</sub>

**Pull request merged**